File: src/models/game_map.py
from .room import Room
from .shrine import Shrine
from .item import Item
from .guardians import *


# 0 -> Empty space
# 1 -> Tunnel
# 2 -> Multip
# 3 -> Divid
# 4 -> Square
# 5 -> Radical
# 6 -> Artifact
room_template = [
    ["Entrance", "Return here with the mathematical artifact. Use it to escape this maze!"],
    ["Tunnel", "Nothing extravagant. An Empty tunnel."],
    ["Multip Shrine", "The Shrine of the Multip sign, granting the power of multiplication.", [
        Item("Multip")], MultipGuardian()],
    ["Divid Shrine", "The Shrine of the Divid sign, granting the power of division.", [
        Item("Divid")], DividGuardian()],
    ["Square Shrine", "The Shrine of the Square sign, granting the power of squaring.", [
        Item("Square")], SquareGuardian()],
    ["Radical Shrine", "The Shrine of the Radical sign, granting the power of square and cube rooting.", [
        Item("Radical")], RadicalGuardian()],
    ["Artifact Shrine", "The Shrine of the Ancient Mathematical Artifact! Grab it, and get out of here!", [
        Item("Artifact")], ArtifactGuardian()]
]


class GameMap:

  def __init__(self):
    self.map = self.create_map()
    self.entrance = self.map[4][1]

  def create_map(self):
    """
    Creates a map like this:
    [S]      [ ][ ]
    [ ][ ][ ][ ][S]
    [ ]   [S]   [ ]
    [ ][S]      [ ]
    [ ][X]   [A][ ]
    """
    # 0 -> Empty space
    # 1 -> Tunnel
    # 2 -> Multip
    # 3 -> Divid
    # 4 -> Square
    # 5 -> Radical
    # 6 -> Artifact
    # 7 -> Entrance
    empty_array = [
        [2, 0, 0, 1, 1],
        [1, 1, 4, 1, 3],
        [1, 0, 1, 0, 1],
        [1, 5, 0, 0, 1],
        [1, 7, 0, 6, 1]
    ]
    return [self.create_rooms(arr) for arr in empty_array]

  def create_rooms(self, room_arr):
    """Creates a room for each number value from the given array"""
    rooms = []
    for room_id in room_arr:
      if room_id == 0:
        # If id is 0, no room
        rooms.append(None)
      elif room_id == 7:
        # If id is 7, Entrance Room
        rooms.append(Room(*room_template[0]))
      elif room_id == 1:
        # If id is 1, Tunnel
        rooms.append(Room(*room_template[1]))
      else:
        # Use the values from the room_template to create a new room
        rooms.append(Shrine(*room_template[room_id]))
    return rooms

  def deprecated_display_info(self):
    """
    Displays Player info and Room info,
    formatted for the utils.display_screen function.

    Returns an array containing:
    - The map
    - An empty string (No 'Info1' section)
    - Player information
    - Room information
    """
    return [
        self.display_map(),
        None,
        self.player,
        self.get_room()
    ]

File: JS/src/models/GameMap.ts
import { Room, RoomType, Shrine } from "./Room";
import { Item } from "./Item";
import {
  MultipGuardian,
  DividGuardian,
  SquareGuardian,
  RadicalGuardian,
  ArtifactGuardian,
  Guardian,
} from "./Guardians";
import { PositionVector2 } from "./PositionVector2";

// 0 -> Entrance
// 1 -> Tunnel
// 2 -> Multip
// 3 -> Divid
// 4 -> Square
// 5 -> Radical
// 6 -> Artifact
const MAP_TEMPLATE: number[][] = [
  [2, 7, 7, 1, 1],
  [1, 1, 1, 1, 3],
  [1, 7, 4, 7, 1],
  [1, 5, 7, 7, 1],
  [1, 0, 7, 6, 1],
];

interface RoomTemplate {
  name: string;
  description: string;
  roomType: RoomType;
  inventory?: Item[];
  guardian?: Guardian;
}

const ROOM_TEMPLATES: RoomTemplate[] = [
  {
    name: "Entrance",
    description: "Return here with the mathematical artifact. Use it to escape this maze!",
    roomType: RoomType.ENTRANCE,
  },

  {
    name: "Tunnel",
    description: "Nothing extravagant. An Empty tunnel.",
    roomType: RoomType.TUNNEL,
  },

  {
    name: "Multip Shrine",
    description: "The Shrine of the Multip sign, granting the power of multiplication.",
    roomType: RoomType.SHRINE,
    inventory: [new Item("Multip")],
    guardian: new MultipGuardian(),
  },
  {
    name: "Divid Shrine",
    description: "The Shrine of the Divid sign, granting the power of division.",
    roomType: RoomType.SHRINE,
    inventory: [new Item("Divid")],
    guardian: new DividGuardian(),
  },
  {
    name: "Square Shrine",
    description: "The Shrine of the Square sign, granting the power of squaring.",
    roomType: RoomType.SHRINE,
    inventory: [new Item("Square")],
    guardian: new SquareGuardian(),
  },
  {
    name: "Radical Shrine",
    description: "The Shrine of the Radical sign, granting the power of square and cube rooting.",
    roomType: RoomType.SHRINE,
    inventory: [new Item("Radical")],
    guardian: new RadicalGuardian(),
  },
  {
    name: "Artifact Shrine",
    description: "The Shrine of the Ancient Mathematical Artifact! Grab it, and get out of here!",
    roomType: RoomType.ARTIFACT,
    inventory: [new Item("Artifact")],
    guardian: new ArtifactGuardian(),
  },
];

export const DEBUG_MAP = [
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
];

/**
 * Creates a map like this:
 * ```
 * [S]      [ ][ ]
 * [ ][ ][ ][ ][S]
 * [ ]   [S]   [ ]
 * [ ][S]      [ ]
 * [ ][X]   [A][ ]
 * ```
 * - S = shrine
 * - A = Artifact Room
 * - X = Cave Entrance.
 */
export class GameMap {
  map: (Room | null)[][];
  entrance: Room;

  constructor(template: number[][] = MAP_TEMPLATE, entraceRoomPosition: PositionVector2 = new PositionVector2(1, 4)) {
    this.map = this.createMap(template);
    this.entrance = this.map[entraceRoomPosition.y][entraceRoomPosition.x]!;
  }

  private createMap = (template: number[][]) => {
    return template.map((rowOfRooms: number[]) =>
      rowOfRooms.map((roomId: number) => {
        const roomInfo = ROOM_TEMPLATES[roomId];

        if (!roomInfo) return null;

        // If there is something in index three, there's a guardian
        // Ergo, we must make a shrine instead of a room
        if (roomInfo.guardian) {
          return new Shrine(roomInfo.name, roomInfo.description, roomInfo.guardian, roomInfo.inventory);
        } else return new Room(roomInfo.name, roomInfo.description, roomInfo.roomType, roomInfo.inventory);
      }),
    );
  };

  public getRoomAtPosition(roomPosition: PositionVector2): Room | null {
    return this.map[roomPosition.y]?.[roomPosition.x] ?? null;
  }
}
